Subject: [PATCH] revision proof
---
Index: packages/plugins/Draw/src/store/reviseFeatures/index.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/plugins/Draw/src/store/reviseFeatures/index.ts b/packages/plugins/Draw/src/store/reviseFeatures/index.ts
--- a/packages/plugins/Draw/src/store/reviseFeatures/index.ts	(revision 447bc95cf5672e21d25432b3d67c844ddd60be01)
+++ b/packages/plugins/Draw/src/store/reviseFeatures/index.ts	(date 1744117758337)
@@ -21,12 +21,17 @@
   }
   const thisController = (abortController = new AbortController())
 
+  const { revision } = getters.configuration
+  if (!revision) {
+    return
+  }
+
   // clone to prevent accidentally messing with the draw tool's data
   let revisedFeatureCollection = cloneFeatureCollection(
     rootGetters['plugin/draw/featureCollection']
   )
 
-  if (getters.configuration.revision?.autofix) {
+  if (revision.autofix) {
     try {
       revisedFeatureCollection = autofixFeatureCollection(
         revisedFeatureCollection
@@ -40,16 +45,16 @@
     }
   }
 
-  if (getters.configuration.revision?.validate) {
+  if (revision.validate) {
     revisedFeatureCollection = validateGeoJson(revisedFeatureCollection)
   }
 
-  if (getters.configuration.revision?.metaServices?.length) {
+  if (revision.metaServices?.length) {
     try {
       revisedFeatureCollection.features = await enrichWithMetaServices(
         revisedFeatureCollection,
         rootGetters.map,
-        getters.configuration.revision.metaServices,
+        revision.metaServices,
         abortController.signal
       )
     } catch (e) {
